key-exchange/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+mod follower;
+mod leader;
+mod msg;
+pub mod point_addition;
+
+use async_trait::async_trait;
+use p256::PublicKey;
+
+use utils_aio::Channel;
+
+pub use follower::KeyExchangeFollower;
+pub use leader::KeyExchangeLeader;
+pub use msg::KeyExchangeMessage;
+pub use point_addition::{PointAddition, PointAdditionError};
+
+pub type KeyExchangeChannel = Box<dyn Channel<KeyExchangeMessage, Error = std::io::Error> + Send>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum KeyExchangeError {
+    #[error("KeyError: {0}")]
+    KeyError(String),
+    #[error("KeyParseError: {0}")]
+    KeyParseError(String),
+    #[error("IOError: {0}")]
+    IOError(#[from] std::io::Error),
+    #[error("UnexpectedMessage: {0:?}")]
+    UnexpectedMessage(KeyExchangeMessage),
+    #[error("PointAdditionError: {0}")]
+    PointAdditionError(#[from] point_addition::PointAdditionError),
+}
+
+#[async_trait]
+pub trait KeyExchange {
+    /// Returns the client's public key share.
+    async fn get_client_key_share(&mut self) -> Result<PublicKey, KeyExchangeError>;
+    /// Sets the server's public key share.
```

Comment:
i cant find "key share" in this rfc https://www.ietf.org/rfc/rfc5246.txt

Reply:
ok, understood. I always think about the result of DH which each party has as "secret key share". But with this new TLS terminology not the result but the input to the DH is called "key share".
This is confusing on its own.  
The confusion is compounded since we also have another "key share" in our protocol: our parties have a share of the ephemeral public key.